extensions/barrage/src/main/java/io/deephaven/extensions/barrage/util/BarrageStreamReader.java

```diff
@@ -219,30 +237,79 @@ public BarrageMessage safelyParseFrom(final StreamReaderOptions options,
                     }
                     final TLongIterator bufferInfoIter = bufferInfo.iterator();
 
-                    final boolean isAddBatch = numAddBatchesRemaining > 0;
-                    if (isAddBatch) {
-                        --numAddBatchesRemaining;
-                    } else {
-                        --numModBatchesRemaining;
-                    }
+                    // add and mod rows are never combined in a batch. all added rows must be received before the first
+                    // mod rows will be received
 
-                    if (isAddBatch) {
-                        final int numRowsTotal = msg.rowsIncluded.intSize("BarrageStreamReader");
+                    if (numAddRowsRead < numAddRowsTotal || (numAddRowsTotal == 0 && numModRowsTotal == 0)) {
                         for (int ci = 0; ci < msg.addColumnData.length; ++ci) {
-                            msg.addColumnData[ci].data = ChunkInputStreamGenerator.extractChunkFromInputStream(options,
-                                    columnChunkTypes[ci], columnTypes[ci], componentTypes[ci], fieldNodeIter,
-                                    bufferInfoIter, ois, (WritableChunk<Values>) msg.addColumnData[ci].data,
-                                    numAddRowsRead, numRowsTotal);
+                            final BarrageMessage.AddColumnData acd = msg.addColumnData[ci];
+
+                            int lastChunkIndex = acd.data.size() - 1;
+
+                            // need to add the batch row data to the column chunks
+                            WritableChunk<Values> chunk = (WritableChunk<Values>) acd.data.get(lastChunkIndex);
+                            int chunkSize = chunk == null ? 0 : chunk.size();
+
+                            final int chunkOffset;
+                            long rowOffset = numAddRowsRead - lastAddStartIndex;
+                            // reading the rows from this batch might overflow the existing chunk
+                            if (rowOffset + batch.length() > chunkSize) {
```

Comment:
If you are intending to overwrite the null, then this if-block should not create a new chunk, right? Perhaps you shouldn't add the null chunk and instead allow this block to construct the initial chunk when numAddRowsTotal is zero.

We need to figure out how to get code coverage on this logic. Can you check if there is a way to clear metadata before uploading to the DoPut in the FlightRoundTrip tests?